File: benchmarks/problems/maximum-subarray/maximum_subarray.c
#include <stdio.h>
#include <stdlib.h>

#define N 1000000L
#define ROUNDS 200

static long long lcg(long long x) {
  return (1103515245LL * x + 12345LL) % 2147483648LL;
}

static void fill(int *a, long long n) {
  long long s = 42;
  for (long long i = 0; i < n; i++) {
    a[i] = (int)(s % 101) - 50;
    s = lcg(s);
  }
}

static long long kadane(const int *a, long long n) {
  long long best = -1000000000LL, cur = 0;
  for (long long i = 0; i < n; i++) {
    long long v = a[i];
    cur = (cur > 0) ? cur + v : v;
    if (cur > best)
      best = cur;
  }
  return best;
}

static long long solve(const int *a, long long n, int rounds) {
  long long checksum = 0;
  for (int r = 0; r < rounds; r++)
    checksum += kadane(a, n);
  return checksum;
}

int main(void) {
  int *a = malloc(N * sizeof(int));
  if (!a)
    return 1;
  fill(a, N);
  long long checksum = solve(a, N, ROUNDS);
  printf("checksum=%lld\n", checksum);
  free(a);
  return 0;
}

File: benchmarks/problems/maximum-subarray/maximum_subarray.rs
fn lcg(x: i64) -> i64 {
    (1103515245i64.wrapping_mul(x) + 12345) % 2147483648
}

fn fill(a: &mut [i32], n: usize) {
    let mut s: i64 = 42;
    for i in 0..n {
        a[i] = (s % 101 - 50) as i32;
        s = lcg(s);
    }
}

fn kadane(a: &[i32], n: usize) -> i64 {
    let mut best: i64 = -1000000000;
    let mut cur: i64 = 0;
    for i in 0..n {
        let v = a[i] as i64;
        cur = if cur > 0 { cur + v } else { v };
        if cur > best {
            best = cur;
        }
    }
    best
}

fn solve(a: &[i32], n: usize, rounds: usize) -> i64 {
    let mut checksum: i64 = 0;
    for _ in 0..rounds {
        checksum += kadane(a, n);
    }
    checksum
}

fn main() {
    let n: usize = 1000000;
    let rounds: usize = 200;
    let mut a = vec![0i32; n];
    fill(&mut a, n);
    let checksum = solve(&a, n, rounds);
    println!("checksum={}", checksum);
}
